Metrics/Microsoft.ApplicationInsights.Metrics.Core/Internal/Microsoft.ApplicationInsights.Metrics/MetricAggregationManager.cs

```diff
@@ -0,0 +1,261 @@
+            }
+            else
+            {
+                AggregatorCollection nextAggregators = new AggregatorCollection(tactTimestamp, futureFilter);
+                prevAggregators = Interlocked.Exchange(ref aggregators, nextAggregators);
+            }
+
+            List<MetricAggregate> persistentValsAggregations = GetPersistentAggregations(tactTimestamp, prevAggregators?.Filter);
+            List<MetricAggregate> nonpersistentAggregations = GetNonpersistentAggregations(tactTimestamp, prevAggregators);
+
+            var summary = new AggregationPeriodSummary(persistentValsAggregations, nonpersistentAggregations);
+            return summary;
+        }
+
+        private List<MetricAggregate> GetPersistentAggregations(DateTimeOffset tactTimestamp, IMetricSeriesFilter previousFilter)
+        {
+            // Complete each persistent aggregator:
+            // The Enumerator of GrowingCollection is a thread-safe lock-free implementation that operates on a "snapshot" of a collection taken at the
+            // time when the enumerator is created. We expand the foreach statement (like the compiler normally does) so that we can use the typed
+            // enumerator's Count property which is constsent with the data in the snapshot.
+
+            GrowingCollection<IMetricSeriesAggregator>.Enumerator persistentValsAggregators = _aggregatorsForPersistent.Aggregators.GetEnumerator();
+            List<MetricAggregate> persistentValsAggregations = new List<MetricAggregate>(capacity: persistentValsAggregators.Count);
+            try
+            {
+                while (persistentValsAggregators.MoveNext())
+                {
+                    IMetricSeriesAggregator aggregator = persistentValsAggregators.Current;
+                    if (aggregator != null)
+                    {
+                        // Persistent aggregators are always active, regardless of filters for a particular cycle.
+                        // But we can apply the cycle's filters to determine whether or not to pull the aggregator
+                        // for a aggregate at this time. Of course, only series filters, not value filters, can be considered.
+                        IMetricValueFilter unusedValueFilter;
+                        bool satisfiesFilter = Util.FilterWillConsume(previousFilter, aggregator.DataSeries, out unusedValueFilter);
+                        if (satisfiesFilter)
+                        {
+                            MetricAggregate aggregate = aggregator.CompleteAggregation(tactTimestamp);
+
+                            if (aggregate != null)
+                            {
+                                persistentValsAggregations.Add(aggregate);
+                            }
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                persistentValsAggregators.Dispose();
```

Comment:
Do we control that GetPersistenctAggregation is called only once, so that we don't Dispose while the other thread is in the while loop? (May it happen in case of double initialization, for instance, we used to have it before?)

Reply:
The thing being disposed in a local object so that it OK in this case.

What did you mean by "Couple places with GetAllPoints() returning vectors"?

Reply:
I remember seeing couple methods named like `GetAllPoints()` that actually returned `vector` object, not `point` object. Purely a naming thing, mathematically you'd use the similar set of numbers to describe vector as to describe point, so you may safely disregard, I suppose.